packages/async-flow/src/async-flow.js

```diff
@@ -0,0 +1,417 @@
+            const { flow } = facets;
+            const eph = tmp.for(flow);
+
+            if (failures.has(flow)) {
+              failures.set(flow, fatalProblem);
+            } else {
+              failures.init(flow, fatalProblem);
+            }
+
+            if (eph.membrane) {
+              eph.membrane.stop();
+            }
+            tmp.resetFor(flow);
+            log.reset();
+            bijection.reset();
+
+            // This is not an expected throw, so in theory arbitrary chaos
+            // may ensue from throwing it. But at this point
+            // we should have successfully isolated this activation from
+            // having any observable effects on the host, aside from
+            // console logging and
+            // resource exhaustion, including infinite loops
+            const err = makeError(
+              X`In a replay failure: see getFailures() for more information`,
+            );
+            annotateError(err, X`due to ${fatalProblem}`);
+            throw err;
+          },
+        },
+        wakeWatcher: {
+          onFulfilled(_fulfillment) {
+            const { facets } = this;
+            facets.flow.wake();
+          },
+          onRejected(_fulfillment) {
+            const { facets } = this;
+            facets.flow.wake();
+          },
+        },
+      },
+    );
+    const makeAsyncFlowKit = (activationThis, activationArgs) => {
+      const asyncFlowKit = internalMakeAsyncFlowKit(
+        activationThis,
+        activationArgs,
+      );
+      asyncFlowKit.flow.restart();
+      return asyncFlowKit;
+    };
+    return harden(makeAsyncFlowKit);
+  };
+
+  /**
+   * @param {Zone} zone
+   * @param {string} tag
+   * @param {GuestAsyncFunc} guestFunc
+   * @param {{ startEager?: boolean }} [options]
+   * @returns {HostAsyncFuncWrapper}
+   */
+  const asyncFlow = (zone, tag, guestFunc, options = undefined) => {
+    const makeAsyncFlowKit = prepareAsyncFlowKit(zone, tag, guestFunc, options);
+    const hostFuncName = `${guestFunc.name || 'anon'}_hostWrapper`;
+    const wrapperFunc = {
+      [hostFuncName](...args) {
+        const { flow } = makeAsyncFlowKit(this, args);
+        const outcomeVow = flow.getOutcome();
+        flowForOutcomeVowKey.init(vowishKey(outcomeVow), flow);
+        return outcomeVow;
+      },
+    }[hostFuncName];
+    defineProperties(wrapperFunc, {
+      length: { value: guestFunc.length },
+    });
+    return harden(wrapperFunc);
+  };
+
+  const adminAsyncFlow = outerZone.exo('AdminAsyncFlow', AdminAsyncFlowI, {
+    getFailures() {
+      return failures.snapshot();
+    },
+    wakeAll() {
+      // [...stuff.keys()] in order to snapshot before iterating
+      const failuresToRestart = [...failures.keys()];
+      const flowsToWake = [...eagerWakers.keys()];
+      for (const flow of failuresToRestart) {
+        flow.restart();
+      }
+      for (const flow of flowsToWake) {
+        flow.wake();
+      }
```

Comment:
Great question! It turns out that the first answer is yes. And in fact, when I did the https://github.com/Agoric/agoric-sdk/pull/9097#discussion_r1581742043 experiment of deleting the mystery test, I found that when the problem occurred during goodReplay, the flow was in both and the problem happened during the call to `flow.wake()`. Together with this question, it had me looking there for an answer for awhile. But that was not the problem, as explained there.

Turns out you are correct, but not quite as you phrased it. If the flow is in both, `restart()` will clear the failure state, so normally by the time `wake()` happens, it will no longer have failed. But it could have failed again later in that same `restart()` before the `wake()` in which case this loop would call `wake()` on a failed flow.

`wake()` should be idempotent. But in this failure state, `wake()` does
```js
            if (failures.has(flow)) {
              throw failures.get(flow);
            }
```
which isn't technically a non-idempotence of `wake` itself. But it does cause the loop above to exit early for no good reason. Changing to 
```js
            if (failures.has(flow)) {
              return;
            }
```
which worked. But changed again to
```js
            const flowState = flow.getFlowState();
            if (flowState === 'Done' || flowState === 'Failed') {
              return;
            }
```
which also works, with more consistency checking.

Reply:
So `flow.wake()` should be now be usefully idempotent if `membrane.wake()` is idempotent. I believe it is, so closing. But if I missed something, please reopen.